lib/src/authentication/token_store.dart

```diff
@@ -0,0 +1,156 @@
+
+/// Stores and manages authentication credentials.
+class TokenStore {
+  TokenStore(this.cacheRootDir);
+
+  /// Cache directory.
+  final String cacheRootDir;
+
+  List<Token>? _tokens;
+
+  /// List of saved authentication tokens.
+  ///
+  /// Modifying this field will not write changes to the disk. You have to call
+  /// [flush] to save changes.
+  List<Token> get tokens => _tokens ??= _loadTokens();
+
+  /// Reads "tokens.json" and parses / deserializes it into list of
+  /// [Token].
+  List<Token> _loadTokens() {
+    final result = List<Token>.empty(growable: true);
+    final path = _tokensFile;
+    if (!fileExists(path)) {
+      return result;
+    }
+
+    try {
+      final json = jsonDecode(readTextFile(path));
+
+      if (json is! Map<String, dynamic>) {
+        throw FormatException('JSON contents is corrupted or not supported.');
+      }
+      if (json['version'] != 1) {
+        throw FormatException('Version is not supported.');
+      }
+
+      if (json.containsKey('hosted')) {
+        if (json['hosted'] is! List) {
+          throw FormatException(
+              'tokens.json format is invalid or not supported.');
+        }
+
+        result.addAll((json['hosted'] as List)
+            .cast<Map<String, dynamic>>()
+            .map((it) => Token.fromJson(it)));
+      }
+    } on FormatException catch (error, stackTrace) {
+      log.error('Failed to load tokens.json', error, stackTrace);
+
+      // When an invalid, damaged or not compatible version of token.json is
+      // found, we remove it after showing error message. Otherwise the error
+      // message will be displayed on each pub command.
+      // Or instead we could write instructrions to execute
+      //`pub token remove --all` if user couldn't solve the issue.
+      deleteEntry(_tokensFile);
+    }
+
+    return result;
+  }
+
+  /// Writes [tokens] into "tokens.json".
+  void _saveTokens(List<Token> tokens) {
+    writeTextFile(
+        _tokensFile,
+        jsonEncode(<String, dynamic>{
+          'version': 1,
+          'hosted': tokens.map((it) => it.toJson()).toList(),
+        }));
+  }
+
+  /// Writes latest state of the store to disk.
+  void flush() {
+    if (_tokens == null) {
+      throw Exception('Schemes should be loaded before saving.');
+    }
+    _saveTokens(_tokens!);
+  }
+
+  /// Adds [token] into store and writes into disk.
+  void addToken(Token token) {
+    // Remove duplicate tokens
+    tokens.removeWhere((it) => it.url == token.url);
+    tokens.add(token);
+    flush();
+  }
+
+  /// Removes tokens with matching [url] from store. Returns whether or not
+  /// there's a stored token with matching url.
+  bool removeMatchingTokens(String url) {
+    final normalizedUrl = validateAndNormalizeHostedUrl(url);
+
+    var i = 0;
+    var found = false;
+    while (i < tokens.length) {
+      if (tokens[i].url == normalizedUrl) {
+        tokens.removeAt(i);
+        found = true;
+      } else {
+        i++;
```

Comment:
removeAt removes current item so in next iteration the item at `i` will be item at `i+1` of previous iteration. It's hard to explain 😅 but [1,2,3] when we remove 1st item from this list index of the next item will become 1.

The code looks a bit weird because it's usually not allowed to modify list while iterating using foreach loop. I'm not sure if dart allows this case or not, at least other languages doesn't allows so I decided to do in this way.